porch/controllers/packagevariants/pkg/controllers/packagevariant/packagevariant_controller.go

```diff
@@ -851,6 +856,122 @@ func getFileKubeObject(prr *porchapi.PackageRevisionResources, file, kind, name
 	return ko, nil
 }
 
+// ensureKRMFunctions adds mutators and validators specified in the PackageVariant to the kptfile inside the PackageRevisionResources.
+// It generates a unique name that identifies the func (see func generatePVFuncname) and moves it to the top of the mutator sequence.
+// It will preserve indent-style if there are no new mutators/validators.
+func ensureKRMFunctions(pv *api.PackageVariant,
+	prr *porchapi.PackageRevisionResources) error {
+
+	if pv.Spec.Pipeline.IsEmpty() {
+		return nil
+	}
+
+	// parse kptfile
+	kptfile, err := getFileKubeObject(prr, kptfilev1.KptFileName, "", "")
+	if err != nil {
+		return err
+	}
+
+	pipeline := kptfile.UpsertMap("pipeline")
+
+	fieldlist := map[string][]kptfilev1.Function{
+		"validators": pv.Spec.Pipeline.Validators,
+		"mutators":   pv.Spec.Pipeline.Mutators,
+	}
+	for fieldname, field := range fieldlist {
+		var newFieldVal = fn.SliceSubObjects{}
+
+		existingFields, ok, err := pipeline.NestedSlice(fieldname)
+		if err != nil {
+			return err
+		}
+		if !ok || existingFields == nil {
+			existingFields = fn.SliceSubObjects{}
+		}
+
+		for _, existingField := range existingFields {
+			ok, err := isPackageVariantFunc(existingField, pv.ObjectMeta.Name)
+			if err != nil {
+				return err
+			}
+			if !ok {
+				newFieldVal = append(newFieldVal, existingField)
+			}
+		}
+
+		var newPVFieldVal = fn.SliceSubObjects{}
+		for i, newFields := range field {
+			newFieldVal := newFields.DeepCopy()
+			newFieldVal.Name = generatePVFuncName(newFields.Name, pv.ObjectMeta.Name, i)
+			f, err := fn.NewFromTypedObject(newFieldVal)
+			if err != nil {
+				return err
+			}
+			newPVFieldVal = append(newPVFieldVal, &f.SubObject)
+		}
+
+		newFieldVal = append(newPVFieldVal, newFieldVal...)
+
+		// if there are new mutators/validators, set them. Otherwise delete the field. This avoids ugly dangling `mutators: []` fields in the final kptfile
+		if len(newFieldVal) > 0 {
+			if err := pipeline.SetSlice(newFieldVal, fieldname); err != nil {
+				return err
+			}
+		} else {
+			if _, err := pipeline.RemoveNestedField(fieldname); err != nil {
+				return err
+			}
+		}
+	}
+
```

Comment:
If the there are no mutators or validators after this, then we should remove the "pipeline" map altogether.